Haskell/Blarney/Core/Prim.hs

```diff
@@ -909,3 +919,38 @@ primInfo prim@(RegFileWrite RegFileInfo{ regFileAddrWidth = aw
            , isRoot = True
            , inputs = [("idx", aw), ("data", dw)]
            , outputs = [] }
+
+-- Nets and netlists
+-- =================
+
+-- | 'Net' type representing a 'Netlist' node
+data Net = Net { -- | The 'Net' 's 'Prim'itive
+                 netPrim         :: Prim
+                 -- | The 'Net' 's 'InstId' identifier
+               , netInstId       :: InstId
+                 -- | The 'Net' 's list of 'NetInput' inputs
+               , netInputs       :: [NetInput]
+                 -- | The 'Net' 's 'NameHints'
+               , netNameHints    :: NameHints
+               } deriving Show
+
+-- | A 'WireId' uniquely identify a wire with a 'Net''s instance identifier
+--   ('InstId') and an output name ('OutputName')
+type WireId = (InstId, OutputName)
+
+-- | A 'Net''s input ('NetInput') can be:
+--   - a wire, using the 'InputWire' constructor
+--   - a complex expression, using the 'InputTree' constructor
+data NetInput = InputWire WireId
+              | InputTree Prim [NetInput]
+              deriving Show
+
+-- | A 'Netlist', represented as an 'Array InstId Net'
+type Netlist = Array InstId Net
+
+-- | Netlist generator
+data NetlistGenerator =
+  NetlistGenerator { getNetlistGenerator :: IO Netlist }
```

Comment:
I am overall not sure I grasp why this is its own type rather than a synonym for IO Netlist